drafts/generating-css-from-scratch-with-postcss.md

```diff
@@ -0,0 +1,113 @@
+
+Fast forward a couple of hours diving through the PostCSS codebase, reading the API docs and several PostCSS plugins, I had a working solution.
+
+### Generating CSS from scratch
+
+To generate CSS with PostCSS you first need to build up an AST:
+
+```js
+const postcss = require("postcss");
+
+const fontFamily = "My Family";
+const root = postcss.root();
+
+const bodyRule = postcss.rule({ selector: "body" }).append(
+  postcss.decl({
+    prop: "font-family",
+    value: `"${fontFamily}"`
+  })
+);
+
+const fontFace = postcss.atRule({ name: "font-face" }).append([
+   postcss.decl({ prop: "font-family", value: `"${fontFamily}"` }),
+   postcss.decl({ prop: "src", value: "url(./fonts/myfont.woff2)" })
+]);
+
+root.append([
+  fontFace,
+  bodyRule
+]);
+```
+
+The CSS code can then be generated from the CSS with `toString()`, resulting in this:
+
+```css
+@font-face {
+    font-family: "My Family";
+    src: url(./fonts/myfont.woff2)
+}
+body {
+    font-family: "My Family"
+}
+```
+
+A couple of things to notice:
+
+* PostCSS does not provide much help in generating proper declaration values [out of the box](https://github.com/TrySound/postcss-value-parser). For example quoting a `font-family` value in case it contains multiple words is not handled by PostCSS automatically.
+* It's easy to conditionally generate properties, rules, declarations etc. The API follows a typical builder pattern, which makes it easy to conditionally call `append`, or not.
+* The CSS output looks opinionated. There are no semicolons after the last declaration and no newlines between rules.
+
+Fortunately, PostCSS is architected quite well and allows you to provide your own ["Stringifier"](http://api.postcss.org/global.html#stringifier). I didn't find much documentation or guidance on this though, but after a bit of code diving I settled on this:
+
+```js
+const Stringifier = require("postcss/lib/stringifier");
+
+class PrettyStringifier extends Stringifier {
+  static new() {
+    return (node, builder) => {
+      let str = new this(builder);
+      str.stringify(node);
+    };
+  }
+
+  constructor(builder) {
+    super(builder);
+  }
+
+  rule(node) {
+    if (node.prev()) {
+      // Add a newline after a rule, if it's preceded by another rule.
+      this.builder("\n", node);
+    }
+
+    return super.rule(node);
+  }
+
+  decl(node) {
+    return super.decl(node, true /* force semicolon */);
+  }
+}
+```
+
+As you can see, this is inheriting most of the default behaviour except for the behaviour around rules and declarations.
+
+The custom `Stringifier` can be used like this:
+
+```js
+root.toString(PrettyStringifier.new());
+```
+
+### Conclusion
+
+All in all I achieved what I had to do, but it didn't feel like PostCSS was particularly well suited for this use-case as everything in PostCSS is geared towards *transforming* CSS.
```

Comment:
Maybe follow up the conclusion with a solid statement, like "It was a nice experiment, but I wouldn't use it for production"?

Reply:
Well I'm not 100% sure about that, which is why I'm asking for readers insights or alternatives below.